src/include/txn.i

```diff
@@ -786,11 +835,22 @@ static inline int
 __wt_txn_update_check(WT_SESSION_IMPL *session, WT_UPDATE *upd)
 {
 	WT_TXN *txn;
+	uint32_t ignore_prepare_set;
 
 	txn = &session->txn;
+	ignore_prepare_set = F_ISSET(txn, WT_TXN_IGNORE_PREPARE);
+	/*
+	 * Clear the ignore prepare setting of txn, as it is not suppose to
+	 * affect the visibility for update operation
+	 */
+	if (ignore_prepare_set)
+		F_CLR(txn, WT_TXN_IGNORE_PREPARE);
+
 	if (txn->isolation == WT_ISO_SNAPSHOT)
 		while (upd != NULL && !__wt_txn_upd_visible(session, upd)) {
 			if (upd->txnid != WT_TXN_ABORTED) {
+				if (ignore_prepare_set)
+					F_SET(txn, WT_TXN_IGNORE_PREPARE);
```

Comment:
This function has gotten kind of messy, and we're doing work we don't have to do if the transaction isn't a `WT_ISO_SNAPSHOT` configuration. The compiler will probably remove that work, but I'm not sure, given as this is an inline function. Here's a suggested rewrite, discard it if you don't like it.
```
/*
 * __wt_txn_update_check --
 *      Check if the current transaction can update an item.
 */
static inline int
__wt_txn_update_check(WT_SESSION_IMPL *session, WT_UPDATE *upd)
{
        WT_TXN *txn;
        bool ignore_prepare_set;
         
        txn = &session->txn;
        if (txn->isolation != WT_ISO_SNAPSHOT)
                return (0);

        /*
         * Clear the ignore prepare setting of txn, as it is not supposed to
         * affect the visibility for update operations.
         */
        ignore_prepare_set = F_ISSET(txn, WT_TXN_IGNORE_PREPARE);
        F_CLR(txn, WT_TXN_IGNORE_PREPARE);
        for (;
            upd != NULL && !__wt_txn_upd_visible(session, upd);
            upd = upd->next)
                if (upd->txnid != WT_TXN_ABORTED)
                        break;
        if (ignore_prepare_set)
                F_SET(txn, WT_TXN_IGNORE_PREPARE);

        if (upd == NULL)
                return (0);

        WT_STAT_CONN_INCR(session, txn_update_conflict);
        WT_STAT_DATA_INCR(session, txn_update_conflict);
        return (__wt_txn_rollback_required(
            session, "conflict between concurrent operations"));
}
```

Reply:
This exact version is failing for test_prepare01.py, so I have modified some part. I still need to  figure out what is the problem with this version.